packages/pluggableWidgets/pie-doughnut-chart-web/src/hooks/data.ts

```diff
@@ -0,0 +1,111 @@
+import { ValueStatus } from "mendix";
+import { useEffect, useMemo, useState } from "react";
+import { ensure } from "@mendix/pluggable-widgets-tools";
+import Big from "big.js";
+import { compareAsc } from "date-fns";
+import { PieChartContainerProps } from "../../typings/PieChartProps";
+import { ChartProps } from "@mendix/shared-charts/dist/components/Chart";
+
+type PieChartDataSeriesHooks = Pick<
+    PieChartContainerProps,
+    | "chartFormat"
+    | "customSeriesOptions"
+    | "seriesColorAttribute"
+    | "seriesDataSource"
+    | "seriesName"
+    | "seriesSortAttribute"
+    | "seriesSortOrder"
+    | "seriesValueAttribute"
+>;
+
+type LocalPieChartData = {
+    itemName: string | undefined;
+    itemValue: number | undefined;
+    itemColor: string | undefined;
+    itemSortValue: string | boolean | Date | Big | undefined;
+};
+
+export const usePieChartDataSeries = ({
+    chartFormat,
+    customSeriesOptions,
+    seriesColorAttribute,
+    seriesDataSource,
+    seriesName,
+    seriesSortAttribute,
+    seriesSortOrder,
+    seriesValueAttribute
+}: PieChartDataSeriesHooks): ChartProps["data"] => {
+    const [pieChartData, setPieChartData] = useState<LocalPieChartData[]>([]);
+
+    useEffect(() => {
+        if (seriesDataSource.status === ValueStatus.Available && seriesDataSource.items) {
+            const dataSourceItems = seriesDataSource.items.map(dataSourceItem => ({
+                itemName: seriesName.get(dataSourceItem).value,
+                itemValue: ensure(seriesValueAttribute).get(dataSourceItem).value?.toNumber(),
+                itemColor: seriesColorAttribute?.get(dataSourceItem).value,
+                itemSortValue: seriesSortAttribute?.get(dataSourceItem).value
+            }));
+            if (seriesSortAttribute) {
+                dataSourceItems.sort(seriesSortAttributeCompareFn);
+                if (seriesSortOrder === "desc") {
+                    dataSourceItems.reverse();
+                }
+            }
+            setPieChartData(dataSourceItems);
+        }
+    }, [
+        seriesColorAttribute,
+        seriesDataSource,
+        seriesName,
+        seriesSortAttribute,
+        seriesSortOrder,
+        seriesValueAttribute
+    ]);
+
+    return useMemo<ChartProps["data"]>(
+        () => [
+            {
+                customSeriesOptions,
+                hole: chartFormat === "pie" ? 0 : 0.4,
```

Comment:
Is it replaceable with custom json properties?

Reply:
Does it means 40% is the blank area?

Reply:
- Yes it's replaceable. This is part of the custom `data` or `series` json property.  See [this code](https://github.com/mendix/widgets-resources/blob/master/packages/shared/charts/src/components/Chart.tsx#L55-L65) where the different types of configurations are merged. Lower precedence is everything from modeler (so the 0.4 that you're referring), then shared options + theme folder, and lastly inline custom configs.
- Yes it's means that 40% of the area (from inner to outer) will be a hole.